src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource_Async.cs

```diff
@@ -97,41 +97,50 @@ private partial class SuggestedActionsSource : IAsyncSuggestedActionsSource
                     // items should be pushed higher up, and less important items shouldn't take up that much space.
                     var currentActionCount = 0;
 
-                    using var _2 = ArrayBuilder<SuggestedActionSet>.GetInstance(out var lowPrioritySets);
+                    var pendingActionSets = new MultiDictionary<CodeActionRequestPriority, SuggestedActionSet>();
 
                     // Collectors are in priority order.  So just walk them from highest to lowest.
                     foreach (var collector in collectors)
                     {
-                        var priority = TryGetPriority(collector.Priority);
-
-                        if (priority != null)
+                        if (TryGetPriority(collector.Priority) is CodeActionRequestPriority priority)
                         {
+                            // See if there are any items we found previously while calling into higher pri providers
+                            // that returned lower pri items. For example 'add import' is a high pri provider, but can
+                            // produce 'low pri' items (like those that add metadata-references). These can be added
+                            // now.
+                            foreach (var set in pendingActionSets[priority])
+                            {
+                                currentActionCount += set.Actions.Count();
+                                collector.Add(set);
+                            }
+
                             var allSets = GetCodeFixesAndRefactoringsAsync(
                                 state, requestedActionCategories, document,
                                 range, selection,
                                 addOperationScope: _ => null,
-                                priority.Value,
+                                priority,
                                 currentActionCount, cancellationToken).WithCancellation(cancellationToken).ConfigureAwait(false);
 
                             await foreach (var set in allSets)
                             {
-                                if (priority == CodeActionRequestPriority.High && set.Priority == SuggestedActionSetPriority.Low)
+                                // Determine the corresponding lightbulb priority class corresponding to the priority
+                                // group the set says it wants to be in.
+                                var actualSetPriority = set.Priority switch
                                 {
-                                    // if we're processing the high pri bucket, but we get action sets for lower pri
-                                    // groups, then keep track of them and add them in later when we get to that group.
-                                    lowPrioritySets.Add(set);
-                                }
-                                else
+                                    SuggestedActionSetPriority.None => CodeActionRequestPriority.Lowest,
```

Comment:
consider extracting a local function next to `ConvertToSuggestedActionSetPriority` at the buttom of this file

Reply:
I started there, but only had it called here, so i liked that i could see exactly what was happening :)